File: java/src/main/java/org/sammancoaching/SmokeTestStep.java
package org.sammancoaching;

import org.sammancoaching.dependencies.Config;
import org.sammancoaching.dependencies.Logger;
import org.sammancoaching.dependencies.Project;

public record SmokeTestStep(String name, Config config, Logger log) implements PipelineStep {
    @Override
    public PipelineStepResult run(Project project, PipelineStepResult stepResult) {
        boolean stepPassed = false;
        String failureReason = stepResult.failureReason();
        if (stepResult.stepPassed()) {
            switch (project.runSmokeTests()) {
                case NO_TESTS -> {
                    failureReason = "Missing Smoke Tests";
                    log.error(failureReason);
                }
                case PASSING_TESTS -> {
                    stepPassed = true;
                    log.info(name() + " passed");
                }
                case FAILING_TESTS -> {
                    failureReason = name() + " failed";
                    log.error(failureReason);
                }
            }
        }

        return new PipelineStepResult(name(), stepPassed, failureReason);
    }
}

File: csharp/BuildPipeline/SmokeTestStep.cs
namespace UntangledConditionals
{
    public class SmokeTestStep : PipelineStep
    {
        private readonly string _name;
        private readonly Config _config;
        private readonly Logger _log;

        public SmokeTestStep(string name, Config config, Logger log)
        {
            _name = name;
            _config = config;
            _log = log;
        }

        public PipelineStepResult Run(Project project, PipelineStepResult stepResult)
        {
            bool stepPassed = false;
            string failureReason = stepResult.FailureReason;

            if (stepResult.StepPassed)
            {
                switch (project.RunSmokeTests())
                {
                    case TestStatus.NoTests:
                        failureReason = "Missing Smoke Tests";
                        _log.Error(failureReason);
                        break;
                    case TestStatus.PassingTests:
                        stepPassed = true;
                        _log.Info(_name + " passed");
                        break;
                    case TestStatus.FailingTests:
                        failureReason = _name + " failed";
                        _log.Error(failureReason);
                        break;
                }
            }

            return new PipelineStepResult(_name, stepPassed, failureReason);
        }
    }
}

